poetry_partifact_plugin/plugin.py

```diff
@@ -0,0 +1,108 @@
+import os
+
+from cleo.events.console_command_event import ConsoleCommandEvent
+from cleo.events.console_events import COMMAND
+from cleo.events.event_dispatcher import EventDispatcher
+from cleo.io.io import IO
+from partifact.auth_token import get_token
+from partifact.config import Configuration
+from poetry.console.application import Application
+from poetry.console.commands.add import AddCommand
+from poetry.console.commands.install import InstallCommand
+from poetry.console.commands.self.self_command import SelfCommand
+from poetry.plugins.application_plugin import ApplicationPlugin
+from tomlkit import parse as parse_toml
+from tomlkit.exceptions import TOMLKitError
+
+CONFIG_PATH = "./pyproject.toml"
+
+
+class PartifactPlugin(ApplicationPlugin):  # type: ignore
+    name = "PartifactPlugin plugin"
+
+    def activate(self, application: Application) -> None:
+        application.event_dispatcher.add_listener(COMMAND, self._handle_pre_command)
+
+    def _handle_pre_command(
+        self, event: ConsoleCommandEvent, event_name: str, dispatcher: EventDispatcher
+    ) -> None:
+        """Check to authenticate to AWS CodeArtifact before any Install or Add Commands"""
+        command = event.command
+        cleo_io = event.io
+
+        if isinstance(command, SelfCommand):
+            # don't run the plugin for self commands
+            return
+
+        if not any(isinstance(command, t) for t in [InstallCommand, AddCommand]):
+            # Only run the plugin for install and add commands
+            return
+
+        try:
+            parsed_toml = self._read_pyproject_file(file_path=CONFIG_PATH)
+        except Exception as error:
+            cleo_io.write_error_line(
+                f"<error>{self.name} failed to read {CONFIG_PATH} pyproject file: \n{error}</>"
+            )
+            return
+
+        if not self._pyproject_toml_has_codeartifact(parsed_toml=parsed_toml):
+            # only run aws login if codeartifact is found in pyproject.toml
+            return
+
+        self._setup_aws_auth(cleo_io=cleo_io, parsed_toml=parsed_toml)
+
+    def _get_sources(self, parsed_toml: dict) -> list:
+        """Return a list of all tool.poetry.source's in the parsed toml."""
+        return parsed_toml.get("tool", {}).get("poetry", {}).get("source", [])
```

Comment:
Should this fail loudly if `tool.poetry.source` is not configured? I think it's preferable to raise a KeyError if required config is missing.

Reply:
No, this function needs to run in my `self._pyproject_toml_has_codeartifact()` function which checks the pyproject.toml file for .codeartifact. in the source block. 

I don't want this to fail loudly because the desired behavior is that the function will simply do nothing if the `tool.poetry.source` block isn't found.